src/protocol/ipa_prf/prf_sharding/mod.rs

```diff
@@ -602,63 +591,58 @@ where
-) -> Result<BitDecomposed<Replicated<Gf2>>, Error>
+    is_trigger_bit: &Replicated<Boolean>,
+    prev_row_breakdown_key_bits: &Replicated<BK>,
+    cur_row_breakdown_key_bits: &Replicated<BK>,
+) -> Result<Replicated<BK>, Error>
 where
-    C: UpgradedContext<Gf2, Share = Replicated<Gf2>>,
+    C: Context,
+    BK: WeakSharedValue + CustomArray<Element = Boolean> + Field,
 {
-    Ok(BitDecomposed::new(
-        ctx.parallel_join(
-            cur_row_breakdown_key_bits
-                .iter()
-                .zip(prev_row_breakdown_key_bits.iter())
-                .enumerate()
-                .map(|(i, (cur_bit, prev_bit))| {
-                    let c = ctx.narrow(&BitOpStep::from(i));
-                    async move { if_else(c, record_id, is_trigger_bit, prev_bit, cur_bit).await }
-                }),
-        )
-        .await?,
-    ))
+    // expand is_trigger_bit to array
+    let is_trigger_bit_array = Replicated::<BK>::expand(is_trigger_bit);
+
+    if_else(
+        ctx,
+        record_id,
+        &is_trigger_bit_array,
+        prev_row_breakdown_key_bits,
+        cur_row_breakdown_key_bits,
+    )
+    .await
 }
 
 /// Same as above but for timestamps. If `attribution_window_seconds` is `None`, just
 /// return the previous row's timestamp. The bits aren't used but saves some multiplications.
-async fn timestamp_of_most_recent_source_event<C>(
+async fn timestamp_of_most_recent_source_event<C, TS>(
     ctx: C,
     record_id: RecordId,
     attribution_window_seconds: Option<NonZeroU32>,
-    is_trigger_bit: &Replicated<Gf2>,
-    prev_row_timestamp_bits: &BitDecomposed<Replicated<Gf2>>,
-    cur_row_timestamp_bits: &BitDecomposed<Replicated<Gf2>>,
-) -> Result<BitDecomposed<Replicated<Gf2>>, Error>
+    is_trigger_bit: &Replicated<Boolean>,
+    prev_row_timestamp_bits: &Replicated<TS>,
+    cur_row_timestamp_bits: &Replicated<TS>,
+) -> Result<Replicated<TS>, Error>
 where
-    C: UpgradedContext<Gf2, Share = Replicated<Gf2>>,
+    C: Context,
+    TS: WeakSharedValue + CustomArray<Element = Boolean> + Field,
 {
     match attribution_window_seconds {
         None => Ok(prev_row_timestamp_bits.clone()),
         Some(_) => {
-            Ok(BitDecomposed::new(
-                ctx
-                    .parallel_join(
-                        cur_row_timestamp_bits
-                            .iter()
-                            .zip(prev_row_timestamp_bits.iter())
-                            .enumerate()
-                            .map(|(i, (cur_bit, prev_bit))| {
-                                let c = ctx.narrow(&BitOpStep::from(i));
-                                async move {
-                                    if_else(c, record_id, is_trigger_bit, prev_bit, cur_bit).await
-                                }
-                            }),
-                    )
-                    .await?,
-            ))
+            // expand is_trigger_bit to array
```

Comment:
```suggestion
```
If you want a comment here, it would be: we need to replicate is_trigger_bit so that contains as many values as the timestamp.

But really, do we need that?